multidimensional-pod-autoscaler/AEP.md

```diff
@@ -0,0 +1,541 @@
+To generate the horizontal scaling action recommendation, we reuse HPA libraries, integrating with the MPA API code, to reads and updates the MPA objects.
+We integrate vertical and horizontal scaling in a single feedback cycle.
+As an intitial solution, vertical scaling and horizontal scaling is performed alternatively (vertical scaling first).
+Vertical scaling will scale the CPU and memory allocations based on the historical usage; and horizontal scaling will scale the number of replicas based on either CPU utilization or a custom metric.
+In the future, we can consider more complex way of prioritization and conflict resolution.
+The separation of recommendation and actuation allows customized recommender to be used to replace the default recommender.
+For example, users can plug-in their RL-based controller to replace the MPA recommender, receiving measurements from the Metrics Server and modifying the MPA objects directly to give recommendations.
+
+The implementation of the MPA framework (the backend) is based on the existing HPA and VPA codebase so that it only requires minimum code maintenance.
+Reused Codebase References:
+- HPA: https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/podautoscaler
+- VPA: https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler
+
+### MPA API Object
+
+We reuse the CR definitions from the [MultidimPodAutoscaler](https://cloud.google.com/kubernetes-engine/docs/how-to/multidimensional-pod-autoscaling) object developed by Google.
+`MultidimPodAutoscaler` is the configuration for multi-dimensional Pod autoscaling, which automatically manages Pod resources and their count based on historical and real-time resource utilization.
+MultidimPodAutoscaler has two main fields: `spec` and `status`.
+
+#### MPA Object
+
+```
+apiVersion: autoscaling.gke.io/v1beta1
+kind: MultidimPodAutoscaler
+metadata:
+  name: my-autoscaler
+# MultidimPodAutoscalerSpec
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: my-target
+  updatePolicy:
+    updateMode: Auto
+  metrics:
+  - type: Resource
+    resource:
+    # Define the target CPU utilization request here
+    name: cpu
+    target:
+      type: Utilization
+      averageUtilization: target-cpu-util
+  constraints:
+    minReplicas: min-num-replicas
+    maxReplicas: max-num-replicas
+    applicationSLA: value  # customizable SLA for application metrics such as latency or throughput
+  resourcePolicy:
+    containerControlledResources: [ memory, cpu ]  # Added cpu here as well
+    container:
+    - name: '*' # either a literal name, or "*" to match all containers
+                # this is not a general wildcard match
+    # Define boundaries for the memory request here
+      requests:
+        minAllowed:
+          memory: min-allowed-memory
+        maxAllowed:
+          memory: max-allowed-memory
+  # Define the recommender to use here
+  recommenders:
```

Comment:
This is now a list of recommenders, not a single recommender to use as the comment suggests. How does it work to have multiple implementations in use at the same time?

Reply:
So it's actually consistent with the definition in the vertical autoscaler: https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L105

Only the first recommender will be supported initially, but later, when multiples are running, only one recommender’s recommendations are selected, and others run in recommendation-only mode.

Reply:
@pbetkier, this is consistent with the VPA support on alternative recommenders. The first support of single alternative recommender is already added, but the multiple alternative recommenders' support for A/B testing is still in planning. We may introduce that soon.